esp/logging/logginglib/modularlogagent.cpp

```diff
@@ -0,0 +1,1526 @@
+    {
+        response.setStatusCode(-1);
+    }
+}
+
+bool CDelegatingAgent::hasService(LOGServiceType type) const
+{
+    switch (type)
+    {
+    case LGSTGetTransactionSeed: return false;
+    case LGSTGetTransactionID:   return false;
+    case LGSTUpdateLOG:          return m_updateLog && m_updateLog->isEnabled();
+    case LGSTterm:
+        ierrlog(Major, "invalid service request for LGSTterm");
+        break;
+    default:
+        ierrlog(Major, "unexpected service request for %d", int(type));
+        break;
+    }
+    return false;
+}
+
+bool CDelegatingAgent::configureUpdateLog(const IPTree& configuration, const CModuleFactory& factory)
+{
+    static Owned<IPTree> defaultConfiguration(createPTreeFromXMLString(VStringBuffer("<%s><%s/></%s>", moduleAgent, moduleUpdateLog, moduleAgent)));
+    Linked<const IPTree> effectiveConfiguration;
+    const char* updateLogValue = configuration.queryProp(propUpdateLog);
+    bool hasScalar = !isEmptyString(updateLogValue);
+    bool hasObject = configuration.hasProp(moduleUpdateLog);
+    if (hasScalar && hasObject)
+    {
+        uerrlog(Major, "%s and %s cannot be configured at the same time", propUpdateLog, moduleUpdateLog);
+        return false;
+    }
+    if (hasScalar)
+    {
+        TokenDeserializer deserializer;
+        bool applyDefault = false;
+        if (deserializer(updateLogValue, applyDefault) != Deserialization_SUCCESS)
+        {
+            uerrlog(Major, "expected Boolean value for %s; found '%s'", propUpdateLog, updateLogValue);
+            return false;
+        }
+        if (!applyDefault)
+            return true;
+        return createAndConfigure(*defaultConfiguration, moduleUpdateLog, factory, factory.m_updateLogs, m_updateLog, *this);
+    }
+    return createAndConfigure(configuration, moduleUpdateLog, factory, factory.m_updateLogs, m_updateLog, *this);
+}
+
+bool CDelegatingAgent::appendProperties(StringBuffer& str) const
+{
+    bool appended = Base::appendProperties(str);
+    bool first = appended;
+    if (m_updateLog)
+    {
+        if (first)
+        {
+            str.append("; ");
+            first = false;
+        }
+        m_updateLog->toString(str);
+        appended = true;
+    }
+    return appended;
+}
+
+//////////////////// CDelegatingUpdateLog ////////////////////
+
+bool CDelegatingUpdateLog::configure(const IPTree& configuration, const CModuleFactory& factory)
+{
+    bool result = Base::configure(configuration, factory);
+    if (!createAndConfigure(configuration, moduleTarget, factory, factory.m_contentTargets, m_target, *this))
+        result = false;
+    return result;
+}
+
+void CDelegatingUpdateLog::updateLog(const char* updateLogRequest, IEspUpdateLogResponse& response)
+{
+    if (!isEnabled())
+    {
+        StringBuffer msg;
+        msg.appendf("service is disabled");
+        response.setStatusCode(-1);
+        response.setStatusMessage(msg);
+        ierrlog(Major, "%s", msg.str());
```

Comment:
Your trace logging component class will be automatically appending a traceID to identify which module generated the message, correct?

Reply:
CModule::traceOutput prefixes messages with the module's trace ID.

Reply:
ok